tests/failpoints/cases/test_replica_stale_read.rs

```diff
@@ -358,27 +326,97 @@ fn test_new_leader_init_resolver() {
     // There are no lock in the region, the `safe_ts` should keep updating by the new leader,
     // so we can read `key1` with the newest ts
     cluster.must_transfer_leader(1, new_peer(2, 2));
-    peer_client1.must_kv_read_equal(
-        b"key1".to_vec(),
-        b"value1".to_vec(),
-        block_on(pd_client.get_tso()).unwrap().into_inner(),
-    );
+    peer_client1.must_kv_read_equal(b"key1".to_vec(), b"value1".to_vec(), get_tso(&pd_client));
 
     // Prewrite on `key2` but not commit yet
     peer_client2.must_kv_prewrite(
         vec![new_mutation(Op::Put, &b"key2"[..], &b"value1"[..])],
         b"key2".to_vec(),
-        block_on(pd_client.get_tso()).unwrap().into_inner(),
+        get_tso(&pd_client),
     );
 
     // There are locks in the region, the `safe_ts` can't be updated, so we can't read
     // `key1` with the newest ts
     cluster.must_transfer_leader(1, new_peer(1, 1));
-    let resp = peer_client2.kv_read(
-        b"key1".to_vec(),
-        block_on(pd_client.get_tso()).unwrap().into_inner(),
-    );
+    let resp = peer_client2.kv_read(b"key1".to_vec(), get_tso(&pd_client));
     assert!(resp.get_region_error().has_data_is_not_ready());
     // But we can read `key1` with `commit_ts1`
     peer_client2.must_kv_read_equal(b"key1".to_vec(), b"value1".to_vec(), commit_ts1);
 }
+
+// Testing that while applying snapshot the follower should reset its `safe_ts` to 0 and
+// reject incoming stale read request, then resume the `safe_ts` after applying snapshot
+#[test]
+fn test_stale_read_while_applying_snapshot() {
+    let (mut cluster, pd_client, leader_client) =
+        prepare_for_stale_read_before_run(new_peer(1, 1), Some(Box::new(configure_for_snapshot)));
+    let mut follower_client2 = PeerClient::new(&cluster, 1, new_peer(2, 2));
+    follower_client2.ctx.set_stale_read(true);
+
+    let k1_commit_ts = leader_client.must_kv_write(
+        &pd_client,
+        vec![new_mutation(Op::Put, &b"key1"[..], &b"value1"[..])],
+        b"key1".to_vec(),
+    );
+    follower_client2.must_kv_read_equal(b"key1".to_vec(), b"value1".to_vec(), k1_commit_ts);
+
+    // Stop replicate data to follower 2
+    cluster.add_send_filter(CloneFilterFactory(
+        RegionPacketFilter::new(1, 2)
+            .direction(Direction::Recv)
+            .msg_type(MessageType::MsgAppend),
+    ));
```

Comment:
Should the filter added before kv_write in order to guarantee the follower 2 won't recv replicate data?

Reply:
It is expected. It is intended to let follower 2 have `(key1, value2)` so we can testing that follower 2 can't read `key1` because it is applying snapshot, not because it didn't have the data. The `(key2, value1)` is intended to testing that follower 2 can read data that replicated through snapshot.

Reply:
got it.